### PR TITLE
CORE: automatic TLs build discovery

### PR DESCRIPTION
**Pull request by vspetrov**
## What
Automatic TLs discovery during build process.
new (optional) configure parameter --with-tls= to specify list of tls to build.

## Why ?
When 3rd party implements custom TL integration should be entirely independent from open-source upstream UCC cude, including build.

### CODE REVIEW

File: src/components/tl/nccl/configure.m4 [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# Copyright (c) 2021      Mellanox Technologies. All rights reserved.
+# $COPYRIGHT$
+# Additional copyrights may follow
+
+CHECK_NCCL
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/components/tl/nccl/configure.m4]: done

File: config/m4/check_tls.m4 [hunk 2]
```diff
@@ -0,0 +1,22 @@
+#
+# Copyright (C) Mellanox Technologies Ltd. 2021.  ALL RIGHTS RESERVED.
+# See file LICENSE for terms.
+#
+
+AC_DEFUN([CHECK_TLS],[
+    AC_ARG_WITH([tls],
```
Comment (bureddy) [inline config/m4/check_tls.m4:7]: what if transport is specified in "tls" but forgot/missed --with-<tls>

Comment (vspetrov) [reply config/m4/check_tls.m4:7]: i don't understand. what is the configure cmd line you are talking about ?
If "--with-tls" is not provided then it equals to "all". All that are found under src/components/tls

Comment (bureddy) [reply config/m4/check_tls.m4:7]: I mean "--with-tls=ucp,nccl,sharp " specified but not  --with-nccl and --with-sharp

Comment (vspetrov) [reply config/m4/check_tls.m4:7]: first param only defines the TLs that are tried to be compiled. if with-tls has nccl nccl will first run CHECK_NCCL from nccl/configure.m4 and if it is not available then tl/nccl is not compiled. bottom line "--with-nccl" is used inside CHECK_NCCL which in turn only performed if nccl is part of required tls list.

**Review by bureddy — APPROVED** [review]

**Review by Sergei-Lebedev — APPROVED** [review]

**Comment by manjugv** [comment]
@vspetrov 

1. Can you please fill these FAQs? 
https://github.com/openucx/ucc/wiki/FAQ#8-how-to-compile-all-tls
https://github.com/openucx/ucc/wiki/FAQ#9-how-to-compile-a-specific-tl

2. Can you update the copyrights? Also, the copyright in build files and "c" files are different.

**Comment by manjugv** [comment]
@vspetrov Maybe it is worthwhile to port this PR to the release. Thoughts?

**Comment by vspetrov** [comment]
@manjugv i'm not sure about port into v1.0.x. It is not a bugfix and we don't have any explicit request for it. I don't see it is necessary. Regarding copyrights: i've updated some. We probably need to review it separately (do we need to change to NV copyrights?)

**Comment by vspetrov** [comment]
bot:retest

**Comment by manjugv** [comment]
> @manjugv i'm not sure about port into v1.0.x. It is not a bugfix and we don't have any explicit request for it. I don't see it is necessary.

That is fine.

**Pull request merged**